### ISSUE TITLE
[BUG] Crashes on startup (on kitty and alacritty)

### ISSUE THREAD

**Issue by radgeRayden** [aristocratos/btop#10]
**Describe the bug**
Tried to run on pop OS 21.04. Crashes on startup with the following log:
```
/usr/include/c++/11/ranges:540: constexpr std::ranges::iota_view<_Winc, _Bound>::iota_view(std::type_identity_t<_Winc>, st
  3 d::type_identity_t<_Bound>) [with _Winc = int; _Bound = int; std::type_identity_t<_Winc> = int; std::type_identity_t<_Boun
  2 d> = int]: Assertion 'bool(__value <= __bound)' failed.                                                                   
  1 fish: “btop” terminated by signal SIGABRT (Abort)
```
It seems that it doesn't like the kitty terminal, but sometimes it seemed random. Unfortunately I wasn't taking note of everything before but I think it happens on kitty but not eg. on gnome terminal.

**To Reproduce**
Run the executable on the kitty terminal.

**Info (please complete the following information):**
 - btop++ version: 1.0.2
 - Architecture: x86_64
 - Platform: Linux
 - (Linux) Kernel: 5.11.0-7620-generic
 - Terminal used: kitty
 - Font used: Victor Mono

**Additional context**

contents of `~/.config/btop/btop.log`

(try running btop with `--debug` flag if error.log is empty)
```
2021/09/20 (19:15:41) | ===> btop++ v.1.0.2
2021/09/20 (19:15:41) | DEBUG: Starting in DEBUG mode!
2021/09/20 (19:15:41) | INFO: Logger set to DEBUG
2021/09/20 (19:15:41) | DEBUG: Setting LC_ALL=en_US.UTF-8
2021/09/20 (19:15:41) | INFO: Running on /dev/pts/1
```

**Comment by dfuehrer** [comment]
This looks like the error I get when I try to start btop in alacritty (```alacritty -e btop```).  That issue is caused by the terminal not reporting the correct size on startup.
At very small sizes, btop will crash because it will try to create a box with one of its dimensions being 0 (this is right before it checks if the window size is large enough to run).  (This causes an iota assert to fail because the first number is larger than the second number).
I think ideally if it starts in too small of a window it should just display that and wait until the window is an appropriate size like it does if it starts in a large window that is resized.
(I don't have the crashing issue in kitty so this could be unrelated but it does seem to error on the same code).

**Comment by aristocratos** [comment]
@dfuehrer 
Great find!
Have changed to show the warning instead on start and can also toggle boxes with the number keys from the warning screen.

@radgeRayden 
Fixed in latest commits, if you need binaries they will be up later today.

**Comment by radgeRayden** [comment]
Works now, thank you.

**Comment by Shinyzenith** [comment]
Can still reproduce on 1.0.6 with sxhkd and
`alacritty -e btop` @aristocratos

**Comment by Shinyzenith** [comment]
![image](https://user-images.githubusercontent.com/60808802/134637266-7edade14-20be-4833-80f5-d00f34bb8e49.png)

**Comment by aristocratos** [comment]
@Shinyzenith 
Can't reproduce it, `alacritty -e btop` works fine when I run it.
The earlier issue with size is fixed, so don't really have anything to go on why it's not working for you.

**Comment by aristocratos** [comment]
@Shinyzenith 
You could try the following to get some more information about the crash:

1. change the keybind to start `alacritty -e gdb btop`

2. `r` to run, wait for crash and press enter.

3. `bt` to get backtrace.

4. Copy the backtrace and post here.

**Comment by Shinyzenith** [comment]
Hey sorry for the late reply! was 4 am when I got the message. Before I just wanna share a small video, it's not exactly the same error but it's similar, crash on launch with sxhkd.

https://user-images.githubusercontent.com/60808802/134808030-1221b82f-99e6-4bfb-b2eb-a4dea472b162.mp4

**Comment by Shinyzenith** [comment]
As you can see in the video, the crash is kind of a hit or miss. Also while testing the `btop-git` AUR package, I noticed that it did not respect the config file where i clearly stated that i wanted `nord.theme`.

**Comment by Shinyzenith** [comment]
Tested with GNU debugger, I could not get it to crash even once with it.

**Comment by Shinyzenith** [comment]
![Sun-Sep-26-06:09:07-PM-IST-2021](https://user-images.githubusercontent.com/60808802/134808441-672932aa-5f6e-47a6-abac-876b09bf1fe0.png)
Still image of the error.

**Comment by aristocratos** [comment]
@Shinyzenith 
Try with `alacritty -e gdb -ex run btop`

**Comment by StreakyCobra** [comment]
I am experiencing the same issue so I'm posting my report here.

---

**Describe the bug**

I am using a tiling window manager. If I run `btop` from an existing terminal it works perfectly. If I want to open `btop` in a new terminal emulator by using `terminator -e btop` it crashes ¹ with the following error message:

```
/usr/include/c++/11.1.0/bits/stl_deque.h:1441: std::deque<_Tp, _Alloc>::const_reference std::deque<_Tp, _Alloc>::back() const [with _Tp = long long int; _Alloc = std::allocator<long long int>; std::deque<_Tp, _Alloc>::const_reference = const long long int&]: Assertion '!this->empty()' failed.
```

¹ It does not crash every time, more on this at the end of this post.

**To Reproduce**

1. Use a tiling window manager (e.g. `i3`)
2. Check that the tiling window manager is configured so that opening a new window splits the screen in two (either vertically of horizontally)
3. Open a first terminal emulator (e.g. `terminator`)
4. From this terminal emulator open `btop` directly in a new terminal emulator, by running something like `terminator -e btop`
5. Eventually close/reopen `btop` this way a few times until it crashes (more on this at the end of this post)

**Expected behavior**

When opening `btop` in a new terminal emulator within a tiling window manager it is does not crash.

**Screenshots**

I managed to get a screenshot of the terminal emulator with `btop` crashing just before it closes:

![image](https://user-images.githubusercontent.com/1587877/134821612-448de778-e6eb-4b2f-b836-c5dc3419eecf.png)

**Info (please complete the following information):**
 - btop++ version: `1.0.9` (d7b5aa9)
 - Binary: self compiled
 - (If compiled) Compiler and version: `g++ (11.1.0)`
 - Architecture: x86_64
 - Platform: Linux
 - (Linux) Kernel: 5.14.7-arch1-1 (Archlinux)
 - Terminal used: `terminator`
 - Font used: `JetBrains Mono`

**Additional context**

```
2021/09/26 (21:37:49) | ===> btop++ v.1.0.9
2021/09/26 (21:37:49) | DEBUG: Starting in DEBUG mode!
2021/09/26 (21:37:49) | INFO: Logger set to DEBUG
2021/09/26 (21:37:49) | DEBUG: Setting LC_ALL=en_US.UTF-8
2021/09/26 (21:37:49) | INFO: Running on /dev/pts/1
```

**GDB Backtrace**

After a lot of tries, I managed to get a backtrace:

```gdb
#0  0x00000000005bf4e2 in raise ()
#1  0x0000000000418276 in abort ()
#2  0x00000000004346fa in std::__replacement_assert(char const*, int, char const*, char const*) ()
#3  0x00000000004799e3 in std::deque<long long, std::allocator<long long> >::back() const ()
#4  0x00000000004643b4 in Cpu::draw[abi:cxx11](Cpu::cpu_info const&, bool, bool) ()
#5  0x000000000043fd3e in Runner::_runner(void*) ()
#6  0x00000000005bbd59 in start_thread ()
#7  0x0000000000637fa3 in clone ()
```

---

## Workaround

Starting `btop` like this may it work most of the time (but see section below).

```
terminator -x 'sleep 0.5; btop'
```

## Supposition

As explained above running `terminator -e btop` fails *most of the time*, and the workaround fixes it *most of the time* as well. I have experienced a bit with different number of windows and with different tiling setup. I can not find a way to replicate the issue 100% of the time. My feeling right now is that the problem is a "race condition" between the the tiling window manager and the terminal emulator/`btop`.

I would guess something in this direction: if the terminal emulator starts `btop` before the tiling window manager had time to create the window and/or make space for it, then `btop` will crash because of the incorrect dimensions it received. If the start of `btop` happens after the tiling window manager did his job then `btop` works as expected.

The workaround of adding a sleep of 0.5 second delay before running `btop` seems to confirm this supposition. And a quick experimentation tells me that the shorter the sleep, the more frequently it crashes.

**Comment by Shinyzenith** [comment]
> @Shinyzenith
> Try with `alacritty -e gdb -ex run btop`

Alright!

Also I think we should make this into a separate issue as the name of this one is misleading and the beginning is about a totally separate bug. ( for search engine sake)

**Comment by Shinyzenith** [comment]
> @Shinyzenith
> Try with `alacritty -e gdb -ex run btop`

Once again I tried a lot but no matter how much I try it does not crash in gdb XD

**Comment by Shinyzenith** [comment]
I suspect when the window is being spawned btop tries to initialize a property of itself from a property of the window, and since the window isn't completely spawned yet, it crashes.

That would also explain why it dose not occur under gdb, because gdb gives it sufficient time to spawn itself.

**Comment by aristocratos** [comment]
@StreakyCobra @Shinyzenith 
Possible fix in latest commits. Added a wait period for max 1000ms if the terminal sizing looks to be off.

**Comment by StreakyCobra** [comment]
@aristocratos I just tried 9233b3c, the bug is still happening (on my side at least).

**Comment by Shinyzenith** [comment]
Can still reproduce https://github.com/aristocratos/btop/commit/9233b3cffc025e7d2b868438393f60d6f11de07e

**Comment by aristocratos** [comment]
@StreakyCobra @Shinyzenith
Ok, another try with latest commit. Changed the order of some of the resizing events.

**Comment by StreakyCobra** [comment]
@aristocratos 65c62ce does not solve the issue.

By the way I appreciate that you are trying to help us and I'm happy to continue helping as much as I can (by trying your fixes for instance), but if you feel that's an edge case and require too much of your time I'm fine living with the workaround :smiley:

**Comment by aristocratos** [comment]
@StreakyCobra 
> By the way I appreciate that you are trying to help us..

That's no problem, this should be solvable :)

But I'm currently at work, so only doing small adjustments when I'm not busy. But it's great that you can test them. 

Haven't been able to reproduce the issue with `alacritty` myself, might be a specific version? I'm building latest from git.

**Comment by StreakyCobra** [comment]
On my side I'm using [`terminator`](https://github.com/gnome-terminator/terminator), and I am able to reproduce the bug with [`rxvt-unicode`](http://software.schmorp.de/pkg/rxvt-unicode.html). So these ones in addition to `alacritty` makes me believe that it's not related to the terminal emulator.

I'm using `i3` as window manager, I don't know if that can be related to it instead of more generally to tiling window manager‽ I tried to force all `terminator` windows to be floating (non-tiled) and I can **not** reproduce the bug this way, so it seems to be really related to the tiling part.

**Comment by StreakyCobra** [comment]
If this can be helpful: if I make the screen to split horizontally by default, then the window is too small for `btop` and it shows me this:

```
Terminal size too small:
Width = 174 Height = 20

Needed for current config:
Width = 80 Height = 24
```

It stays on this "screen" as I continue to resize it and the values of "height" are updated until I make it `>=24`, and only then it crashes.

**Comment by aristocratos** [comment]
@StreakyCobra 
Moved the check for valid terminal dimensions to earlier in the initialization, might work.
Try with latest commit.

**Comment by StreakyCobra** [comment]
Still no luck with 01dfd27.

Could it be that it gets a valid width and height right from the beginning (for instance the whole screen dimensions), but then the window manager handle the window and split the screen so the dimensions changes, and when the program tries to draw the screen it is still using the previously announced dimensions?

**Comment by aristocratos** [comment]
@StreakyCobra 
Just to make sure, you run `make clean` before recompiling?

**Comment by aristocratos** [comment]
@StreakyCobra 
Pushed another possible fix, if it still crashes now, it should at least give another error message.

**Comment by StreakyCobra** [comment]
> Just to make sure, you run make clean before recompiling?

I'm using the [AUR `btop-git` package](https://aur.archlinux.org/packages/btop-git/) and the AUR build system takes care of this (as far as I know).

But you are right it's probably "safer" to build from source manually. I just did that and the problem is still happening. To be sure everything was successful I changed the `const string Version` to check that compilation is working as expected.

Edit: I'll go try your latest change now

**Comment by StreakyCobra** [comment]
Nope, still the same error message:

```
/usr/include/c++/11.1.0/bits/stl_deque.h:1441: std::deque<_Tp, _Alloc>::const_reference std::deque<_Tp, _Alloc>::back() const [with _Tp = long long int; _Alloc = std::allocator<long long int>; std::deque<_Tp, _Alloc>::const_reference = const long long int&]: Assertion '!this->empty()' failed.
```

with the same stacktrace:

```
#0  0x00000000005bf482 in raise ()
#1  0x00000000004189de in abort ()
#2  0x0000000000433bb8 in std::__replacement_assert(char const*, int, char const*, char const*) ()
#3  0x00000000004928a3 in std::deque<long long, std::allocator<long long> >::back() const ()
#4  0x000000000047f442 in Cpu::draw[abi:cxx11](Cpu::cpu_info const&, bool, bool) ()
#5  0x000000000043e749 in Runner::_runner(void*) ()
#6  0x00000000005bbcf9 in start_thread ()
#7  0x0000000000637f43 in clone ()
```

I am on commit a09ad063672c73743670603c28cd9c68732d5db4, I ran `make clean` and `make STATIC=true` and ensured changes where compiled by modifying the version and checking `btop --version` reports the modified one.

**Comment by Shinyzenith** [comment]
> @StreakyCobra @Shinyzenith
> Ok, another try with latest commit. Changed the order of some of the resizing events.
Can still reproduce on https://github.com/aristocratos/btop/commit/4f3db691a58836c9045507e6e4e570db7d0ff0b3

**Comment by aristocratos** [comment]
@StreakyCobra @Shinyzenith 
I'm pretty sure I know somewhat where this is happening (when the graphs in the CPU box is generated and calling the `back()` function on the list holding either the cpu percent or cpu temperature values.

But I'm gonna have to set up a proper test environment so I can reproduce the exception myself. Just uploading commit on commit with you guys testing it out is gonna be a bit tedious. 

But should have some time to take a proper look at this later this week.

It could be helpful if you both wanted to post a short list of the terminals and window manager combinations you have gotten the exception with so far (to make it easier for me to set up the testing).

**Comment by StreakyCobra** [comment]
For me it is:

- `i3` / `terminator`
- `i3` / `rxvt-unicode`

**Comment by dfuehrer** [comment]
This crash seems to happen at line 609 in btop_draw.cpp for me on calling ```.back()``` for the temp (```cpu.temp.at(0).back()```).  I don't understand the architecture nearly well enough to understand why the queue would be empty here (and especially only sometimes).
System info for me is ```dwm``` and ```alacritty``` on Arch and a Ryzen 4800H (in case this system info is helpful).

**Comment by aristocratos** [comment]
@dfuehrer 
Based on your comment I made a change in one the previous quick fixes.
@StreakyCobra @Shinyzenith @dfuehrer 
Please try with latest commit and see if we're in luck :)

**Comment by aristocratos** [comment]
Also moved the setup of the signal handlers to later in initialization.
I'm thinking the issue could be that the quick resizes at startup triggers the signal handler for SIGWINCH while the first run of collection is still going and in turn stops at the point where cpu temperature information is collected. Leading to the empty deque...

**Comment by dfuehrer** [comment]
This bug is fairly inconsistent for me but it does seem to fix the issue for me (I can still replicate it in previous commits but cannot anymore)

**Comment by aristocratos** [comment]
@dfuehrer 
Awesome to hear, I think the commit i pushed seconds ago (moving the signal handlers), should also make the previous fixes redundant.

**Comment by StreakyCobra** [comment]
I also can't reproduce it any longer :tada:

**Comment by aristocratos** [comment]
Should be fixed as of v1.0.10 but will wait on @Shinyzenith to also confirm before closing this.

**Comment by Shinyzenith** [comment]
Hi sorry for the late reply. Testing latest commit.

**Comment by Shinyzenith** [comment]
You're a legend, can no longer reproduce this.

**Comment by Shinyzenith** [comment]
Hoping this gets pushed to AUR soon, will package the fix into my distribution <3

**Comment by aristocratos** [comment]
Awesome and thanks everybody for the testing the commits and saving me a bunch of hassle with trying to replicate :)
Closing this.
